rmw_connext_dynamic_cpp/src/functions.cpp

```diff
@@ -878,7 +762,59 @@ rmw_destroy_publisher(rmw_node_t * node, rmw_publisher_t * publisher)
     } \
   }
 
-#define SET_STRING_VALUE(TYPE, METHOD_NAME) \
+#define SET_C_STRING_VALUE(TYPE, METHOD_NAME) \
+  { \
+    if (member->is_array_) { \
+      DDS_DynamicData dynamic_data_member(NULL, DDS_DYNAMIC_DATA_PROPERTY_DEFAULT); \
+      DDS_ReturnCode_t status = dynamic_data->bind_complex_member( \
+        dynamic_data_member, \
+        NULL, \
+        static_cast<DDS_DynamicDataMemberId>(i + 1)); \
+      if (status != DDS_RETCODE_OK) { \
+        RMW_SET_ERROR_MSG("failed to bind complex member"); \
+        return false; \
+      } \
+      ARRAY_SIZE_AND_VALUES(TYPE) \
+      if (array_size > (std::numeric_limits<DDS_DynamicDataMemberId>::max)()) { \
+        RMW_SET_ERROR_MSG( \
+          "failed to set string since the requested string length exceeds the DDS type"); \
+        return false; \
+      } \
+      for (size_t j = 0; j < array_size; ++j) { \
+        status = dynamic_data_member.METHOD_NAME( \
+          NULL, \
+          static_cast<DDS_DynamicDataMemberId>(j + 1), \
+          ros_values[j].data); \
+        if (status != DDS_RETCODE_OK) { \
+          RMW_SET_ERROR_MSG("failed to set array value using " #METHOD_NAME); \
+          return false; \
+        } \
+      } \
+      status = dynamic_data->unbind_complex_member(dynamic_data_member); \
+      if (status != DDS_RETCODE_OK) { \
+        RMW_SET_ERROR_MSG("failed to unbind complex member"); \
+        return false; \
+      } \
+    } else { \
+      auto typed_ros_message = static_cast<const TYPE *>(ros_message); \
+      if (!typed_ros_message) { \
+        RMW_SET_ERROR_MSG("Failed to cast ROS message as TYPE"); \
+        return false; \
+      } \
+      const TYPE * value = reinterpret_cast<const TYPE *>( \
+        typed_ros_message + member->offset_); \
+      DDS_ReturnCode_t status = dynamic_data->METHOD_NAME( \
+        NULL, \
+        static_cast<DDS_DynamicDataMemberId>(i + 1), \
+        value->data); \
+      if (status != DDS_RETCODE_OK) { \
+        RMW_SET_ERROR_MSG("failed to set value using " #METHOD_NAME); \
+        return false; \
+      } \
+    } \
+  }
```

Comment:
Since this is a block of pretty complicated code how is this different to `SET_CPP_STRING_VALUE`?